### PR TITLE
EPL-7606: Reorganize JsonSchema Class

### PR DESCRIPTION
**Pull request by michaelcarter-wf**
## Ultimate problem:
- The main `JsonSchema` class is a large file, was pretty disorganized, and many of the functions contained within had a lot of side effects that made flow hard to follow. 

## How it was fixed:
- Refactor some validations into a new file
- Rough alpha / category organization of the file
- Make validation flow more functional

## Testing suggestions:
- [x] CI Passes

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Module Findings
No security relevant content detected. Please review for security relevance and request security review as needed.

**Comment by semveraudit-wf** [comment]
### Public API Changes

 Recommendation: **:warning: Minor version bump**


```diff
@@ line 48: package:json_schema/src/json_schema/json_schema.dart @@
class JsonSchema {}
+     JsonSchema createSchemaWithProvidedRefs(Map data, Map<String, JsonSchema> providedRefs);
//    Adding a method is a minor change.

```

Showing results for 252fbaa0e7d8695c5bb0a7c243a9f98a846a4ab3
> Powered by [semver-audit-service](https://github.com/Workiva/semver-audit-service). Please report any problems by [filing an issue](https://github.com/Workiva/semver-audit-service/issues).
[Browse public API](https://semver-audit.wk-dev.wdesk.org/Workiva/json_schema/252fbaa0e7d8695c5bb0a7c243a9f98a846a4ab3).

**Comment by jamesromo-wf** [comment]
Looks like Travis is angry

**Comment by michaelcarter-wf** [comment]
Will attempt to fix travis when we prepare to release 2.0.0, but not planning to deal with that in this PR

**Comment by jamesromo-wf** [comment]
+1 CI passes and changes look good.

File: lib/src/json_schema/json_schema.dart [hunk 1]
```diff
@@ -58,170 +64,541 @@ class JsonSchema {
...
+  int get maxProperties => _maxProperties;
+
+  /// The minimum number of properties allowed.
+  int get minProperties => _minProperties;
+
+  /// Map of [JsonSchema]s for properties, based on [RegExp]s keys.
   Map<RegExp, JsonSchema> get patternProperties => _patternProperties;
-  Map<String, JsonSchema> get schemaDependencies => _schemaDependencies;
+
+  /// TODO: write an informative comment for this.
```
Comment (jordanross-wf) [inline lib/src/json_schema/json_schema.dart]: Do you want to resolve this TODO before merging this PR?

File: lib/src/json_schema/json_schema.dart [hunk 2]
```diff
@@ -58,170 +64,541 @@ class JsonSchema {
...
+  /// Validate, calulcate and set the value of the 'not' JSON Schema prop.
+  _setNot(dynamic value) => _makeSchema('$_path/not', TypeValidators.object('not', value), (rhs) => _notSchema = rhs);
+
+  /// Validate, calulcate and set the value of the 'oneOf' JSON Schema prop.
+  _setOneOf(dynamic value) => _validateListOfSchema('oneOf', value, (schema) => _oneOf.add(schema));
+
+  /// Validate, calulcate and set the value of the 'pattern' JSON Schema prop.
+  _setPattern(dynamic value) => _pattern = new RegExp(TypeValidators.string('pattern', value));
+
+  /// Validate, calulcate and set the value of the 'ref' JSON Schema prop.
```
Comment (jordanross-wf) [inline lib/src/json_schema/json_schema.dart]: #nit `calulcate` typo on all of these comments :)

File: lib/src/json_schema/type_validators.dart [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      for (int j = i + 1; j < value.length; j++) {
+        if (JsonSchemaUtils.jsonEqual(value[i], value[j]))
+          throw FormatExceptions.error('enum values must be unique: $value [$i]==[$j]');
+      }
+      i++;
+    });
+    return enumValues;
+  }
+
+  /// Validate A dynamic value is a String.
```
Comment (jordanross-wf) [inline lib/src/json_schema/type_validators.dart]: #nit `A` should be lowercase

File: lib/src/json_schema/validator.dart [hunk 4]
```diff
@@ -153,33 +153,34 @@ class Validator {
...
-    var pattern = schema.pattern;
+    final pattern = schema.pattern;
     if (pattern != null && !pattern.hasMatch(instance)) {
       _err('${schema.path}: pattern violated ($instance vs $pattern)');
     }
   }
 
   void _itemsValidation(JsonSchema schema, dynamic instance) {
-    int actual = instance.length;
+    // TODO: throw an error if instance.length doesn't exist.
```
Comment (jordanross-wf) [inline lib/src/json_schema/validator.dart]: Will this be addressed in a future PR?

File: test/unit/browser/json_schema/validation_test.dart [hunk 5]
```diff
@@ -48,7 +48,7 @@ final Logger _logger = new Logger('test_validation');
 void main([List<String> args]) {
   configureJsonSchemaForBrowser();
 
-  if (args?.isEmpty ?? false) {
+  if (args?.isEmpty != true) {
```
Comment (jordanross-wf) [inline test/unit/browser/json_schema/validation_test.dart]: This change reverses the logic that was in place before, is that intended?

File: test/unit/vm/json_schema/invalid_schemas_test.dart [hunk 6]
```diff
@@ -53,37 +53,36 @@ final Logger _logger = new Logger('test_invalid_schemas');
 void main([List<String> args]) {
   configureJsonSchemaForVm();
 
-  if (args?.isEmpty ?? false) {
+  if (args?.isEmpty != true) {
```
Comment (jordanross-wf) [inline test/unit/vm/json_schema/invalid_schemas_test.dart]: Same thing with this, is this intended to change the logic of the conditional?

File: test/unit/vm/json_schema/validation_test.dart [hunk 7]
```diff
@@ -51,7 +51,7 @@ final Logger _logger = new Logger('test_validation');
 void main([List<String> args]) {
   configureJsonSchemaForVm();
 
-  if (args?.isEmpty ?? false) {
+  if (args?.isEmpty != true) {
```
Comment (jordanross-wf) [inline test/unit/vm/json_schema/validation_test.dart]: double checking logic change here too

File: lib/src/json_schema/json_schema.dart (hunk 2 above)
Comment (michaelcarter-wf) [reply lib/src/json_schema/json_schema.dart]: 🤣

File: lib/src/json_schema/json_schema.dart (hunk 1 above)
Comment (michaelcarter-wf) [reply lib/src/json_schema/json_schema.dart]: nope, don't really know what this does :/...

File: test/unit/browser/json_schema/validation_test.dart (hunk 5 above)
Comment (michaelcarter-wf) [reply test/unit/browser/json_schema/validation_test.dart]: good catch!

File: lib/src/json_schema/validator.dart (hunk 4 above)
Comment (michaelcarter-wf) [reply lib/src/json_schema/validator.dart]: addressed
